src/main/java/io/anserini/search/SearchSolrCollection.java

```diff
@@ -0,0 +1,337 @@
+           * the fourth column is the rank the document is retrieved.
+           * the fifth column shows the score (integer or floating point) that generated the ranking.
+           * the sixth column is called the "run tag" and should be a unique identifier for your
+           */
+          for (int i = 0; i < docs.documents.length; i++) {
+            out.println(String.format(Locale.US, "%s Q0 %s %d %f %s", qid,
+                    docs.documents[i].getField(FIELD_ID).stringValue(), (i + 1), docs.scores[i], runTag));
+          }
+        }
+        out.flush();
+        out.close();
+
+        // Needed for orderly shutdown so the SolrClient executor does not delay main thread exit
+        try {
+          solrPool.returnObject(client);
+          solrPool.close();
+        } catch (Exception e) {
+          LOG.error("Exception in SolrClient executor: ", e);
+        }
+
+        final long durationMillis = TimeUnit.MILLISECONDS.convert(System.nanoTime() - start, TimeUnit.NANOSECONDS);
+        LOG.info("[Finished] Run " + topics.size() + " topics searched in "
+                + DurationFormatUtils.formatDuration(durationMillis, "HH:mm:ss"));
+      } catch (Exception e) {
+        LOG.error(Thread.currentThread().getName() + ": Unexpected Exception:", e);
+      }
+    }
+  }
+
+  public SearchSolrCollection(Args args) throws IOException {
+    this.args = args;
+    LOG.info("Solr index: " + args.solrIndex);
+    LOG.info("Solr ZooKeeper URL: " + args.zkUrl);
+    LOG.info("SolrClient pool size: " + args.solrPoolSize);
+    GenericObjectPoolConfig<SolrClient> config = new GenericObjectPoolConfig<>();
+    config.setMaxTotal(args.solrPoolSize);
+    config.setMinIdle(args.solrPoolSize); // To guard against premature discarding of solrClients
+    this.solrPool = new GenericObjectPool(new SolrClientFactory(), config);
+  }
+
+  @SuppressWarnings("unchecked")
+  public<K> void runTopics() throws IOException {
+    TopicReader<K> tr;
+    SortedMap<K, Map<String, String>> topics = new TreeMap<>();
+    for (String singleTopicsFile : args.topics) {
+      Path topicsFilePath = Paths.get(singleTopicsFile);
+      if (!Files.exists(topicsFilePath) || !Files.isRegularFile(topicsFilePath) || !Files.isReadable(topicsFilePath)) {
+        throw new IllegalArgumentException("Topics file : " + topicsFilePath + " does not exist or is not a (readable) file.");
+      }
+      try {
+        tr = (TopicReader<K>) Class.forName("io.anserini.search.topicreader." + args.topicReader + "TopicReader")
+            .getConstructor(Path.class).newInstance(topicsFilePath);
+        topics.putAll(tr.read());
+      } catch (Exception e) {
+        throw new IllegalArgumentException("Unable to load topic reader: " + args.topicReader);
+      }
+    }
+  
+    final String runTag = args.runtag == null ? "Solrini" : args.runtag;
+    final ThreadPoolExecutor executor = (ThreadPoolExecutor) Executors.newFixedThreadPool(args.threads);
```

Comment:
Do we need the complexity of a thread pool here as it's just a single `SolrClient` doing work? I'd remove the `ThreadPool` and the `ObjectPool<SolrClient>` - it's used in indexing for performance, but the querying happens from a single `SolrClient`?